File: gomoserv/js/gomoku.js
var ws;
var mode;
var turn = "null";
var score = "Black : 0 | White : 0";
var scoreb = 0;
var scorew = 0;
var me = "You are : unknown";
var color = "null";
var host = window.location.hostname;
var portWS;

//init page
$("#turn").text("IDK");
$("#score").text(score);
$("#me").text(me);

function getPortWS() {
  var url = window.location.href;
  var arr = url.split("/");
  var result = arr[2].split(":")[1];
  //portWS = parseInt(result, 0);
  portWS = result;
  //console.log(portWS);
}

function getURLParameter(name) {
  return decodeURI(
  (RegExp(name + '=' + '(.+?)(&|$)').exec(location.search) || [null])[1]);
}

$(document).ready(function() {
  var res = getURLParameter("win");
  if(res == "1") {
    $("#victory").text("YOU WON").show(300);
  } else if(res == "2") {
    $("#victory").text("YOU LOSE").show(300);
  }
   $("#mode").hide(200);
   getPortWS();
});

function sendRules() {
  var dual3 = 0;
  var break5 = 0;
  var timeout = 0;

  if($('#DOUBLE_3').attr('checked')) {
    dual3 = "1";
  }
  if($('#BREAKING_5').attr('checked')) {
    break5 = "1";
  }
  timeout = $('#TIMEOUT').val();
  ws.send("RULES " + dual3 + " " + break5 + " " + timeout);
}

function ConnectWS() {

  ws = new WebSocket("ws://" + host + ":"+portWS+"/ws");

  ws.onopen = function() {
    console.log("Connected to server");
    ws.send("MODE " + mode);
    if(color == "null") {
      ws.send("GETCOLOR");
    }
    ws.send("GETTURN");
    $("#sendrules").removeAttr("disabled");
     $("#reset").removeAttr("disabled");
    $(".infoconnect").css("color", "green");
    $(".infoconnect").html("Connected");
  };

  ws.onmessage = function(e) {
    console.log("receive : '" + e.data + "'");

    if(e.data != "error") {
      var data = e.data.split(' ');

      if(data[0] == "COLOR") {
        color = data[1];
      }

      if(data[0] == "RULES") {
        if (data[1] == "1") {
          $('#DOUBLE_3').attr('checked', 'checked');
        } else if (data[1] == "0") {
          $('#DOUBLE_3').removeAttr('checked');
        }
        if (data[2] == "1") {
          $('#BREAKING_5').attr('checked', 'checked');
        } else if (data[2] == "0") {
          $('#BREAKING_5').removeAttr('checked');
        }
        $('#TIMEOUT').val(data[3]);
      }

      if (data[0] == "HINT") {
        $('.boardclic td div').removeClass("hintstone");
        $('.pos' + data[1] + 'y' + data[2]).addClass("hintstone");
      }

      if(data[0] == "COLOR") {
        color = data[1];
        $("#me").text("You are : " + color);
      }

      if(data[0] == "TURN") {
        turn = data[1];
      }
      //he won
      if(data[0] == "LOSE") {
        ws.send("reset");
        window.location.href = location.protocol + '//' + location.host + location.pathname + "?win=2";
      }

      //you won
      if(data[0] == "WIN") {
        ws.send("reset");
        window.location.href = location.protocol + '//' + location.host + location.pathname + "?win=1";
      }

      //stone captured
      else if(data[0] == "REM") {
        if (turn == "black") {
          scoreb += 2;
        } else if (turn == "white") {
          scorew += 2;
        }
        $('.pos' + data[1] + 'y' + data[2]).removeClass("bgblack");
        $('.pos' + data[1] + 'y' + data[2]).removeClass("bgwhite");
        $('.pos' + data[3] + 'y' + data[4]).removeClass("bgblack");
        $('.pos' + data[3] + 'y' + data[4]).removeClass("bgwhite");
        $("#score").text("Black : "+scoreb+" | White : "+scorew);
      }

      else if(data[0] == "YOURTURN") {
        $("#turn").text("YOUR TURN");
      } else if(data[0] == "ADD") {
        $('.boardclic td div').removeClass("hintstone");
        $('.boardclic td div').removeClass("newstone");
        if(turn == color) {
          $("#turn").text("NOT YOU");
        }
        $('.pos' + data[1] + 'y' + data[2]).addClass("bg" + turn);
        $('.pos' + data[1] + 'y' + data[2]).addClass("newstone");
        if(turn == "black") {
          turn = "white";
        } else {
          turn = "black";
        }
      }

    }
  };


  ws.onclose = function(e) {
    console.log("Disconnected from server");
    $(".infoconnect").css("color", "red");
    $(".infoconnect").html("Disconnected ;(");
  };

}

//click gamemode
$(".selectpvp").click(function() {
  mode = "pvp";
  ConnectWS();

  $("#victory").hide();
  $("#menu").slideUp();
  $("#game").show();
});

$(".selectpve").click(function() {
  mode = "pve";
  ConnectWS();
  color = "black";
  turn = "black";
  $("#victory").hide();
  $("#menu").slideUp();
  $("#game").show();
  $('#hint').attr('disabled', 'disabled');
});

//togglesetting

$(".settingmenuswitch").click(function() {
  if($('#mode').css('display') == "none" ) {
    $("#mode").show(200);
  } else {
    $("#mode").hide(200);
  }
});

$("#sendrules").click(function() {
   sendRules();
});

//give player an hint
$("#hint").click(function() {
  ws.send("HINT");
});

//reset all game
$("#reset").click(function() {
  ws.send("reset");
  location.reload();
});

//new move
$(".boardclic td").click(function() {
  var stone = $(this).find(".stone");
  var coord = stone.text();
  ws.send("PLAY " + coord);
  //ws.send("getscore");
});

File: gomoserv/gomoku.go
package main

import (
	"code.google.com/p/go.net/websocket"
	"html/template"
	"log"
	"net"
	"strings"
)

var (
	Board      [][]int = initBoard(GOBANSIZE)
	Turn               = BLACK
	BPOW               = 0
	WPOW               = 0
	Mode               = UNKNOWN
	DOUBLE_3           = 0
	BREAKING_5         = 0
	TIMEOUT            = 0
	players            = make(map[Connection]int)
	tm                 = 0
	RULES_ST           = "RULES 0 0 0"
)

const (
	UNKNOWN         = -1
	PVP             = 1
	PVE             = 2
	BLACK           = 1
	WHITE           = 2
	IA              = 3
	NONE            = 0
	GOBANSIZE       = 20
	NB_ALIGN_TO_WIN = 5
	STONE_TO_Win    = 10
	W_CAPTURE       = 1
	W_FIVEALIGN     = 2
	W_RULEERR       = 3
	W_TIMEOUT       = 4
)

type Page struct {
	Title      string
	Board      template.HTML
	BoardClick template.HTML
}

type Connection struct {
	ws           *websocket.Conn
	s            net.Conn
	player_color int
	is_ia        bool
	clientip     string
}

func engine(msg_cl string, con Connection) int {
	buff := strings.Split(msg_cl, " ")

	buf := buff[0]

	if con.player_color == NONE {
		return 1
	}

	//check le mode
	if Mode == UNKNOWN {
		if buf == "MODE" {
			if buff[1] == "pve" {
				Mode = PVE
				con.player_color = BLACK
			}
			if buff[1] == "pvp" {
				Mode = PVP
			}
		}
		return 1
	}

	//check avec le referee
	switch buf {
	case "reset":
		for pl, _ := range players {
			delete(players, pl)
		}
		players = make(map[Connection]int)
		Board = initBoard(GOBANSIZE)
		DOUBLE_3 = 0
		BREAKING_5 = 0
		TIMEOUT = 0
		return -1

	case "RULES":
		DOUBLE_3 = Atoi(buff[1])
		BREAKING_5 = Atoi(buff[2])
		TIMEOUT = Atoi(buff[3])
		RULES_ST = msg_cl
		for pl, _ := range players {
			send(msg_cl, pl)
		}
	case "GETCOLOR":
		send("COLOR "+getStringPl(con.player_color), con)
	case "GETTURN":
		send("TURN "+getStringTurn(), con)
	case "CONNECT":
		if buff[1] == "CLIENT" {
			log.Print("IA Connected")
			con.is_ia = true
			con.player_color = WHITE
			send(RULES_ST, con)
			if Turn == WHITE {
				send("YOURTURN", con)
			}
		}
	case "PLAY":
		tm++
		coord := []string{buff[1], buff[2]}
		if len(coord) > 1 {
			mov, win, _ := referee(coord, con)
			if win {
				send_win(con, W_FIVEALIGN)
				return -1
			} else if mov == true {
				buf = "ADD " + buff[1] + " " + buff[2]
				for pl, _ := range players {
					send(buf, pl)
					if pl.player_color == Turn {
						send("YOURTURN", pl)
						if Mode == PVE && Turn == WHITE && TIMEOUT > 0 {
							go timeout_to_death()
						}
					}
				}
			} else {
				if Mode == PVE && con.player_color == WHITE {
					for pl, _ := range players {
						if pl.player_color == BLACK {
							send_win(pl, W_RULEERR)
							return -1
						}
					}
				}
			}
			//send(buf, ws)
			AffBoard(Board, GOBANSIZE)
		}

	}
	return 1
}
